### ISSUE TITLE
qt5 port missing qml [related to optional features]

### ISSUE THREAD

**Issue by japj** [microsoft/vcpkg#787]
Hi, I am trying to try and use vcpkg for MuseScore dependencies.
I successfully build qt5 using VS2017.

However, MuseScore uses QML from QT, which is unfortunately disabled in the port file (-skip declarative). As a result the MuseScore CMake detection fails to find QML.

Next it that it also seems that the declarative part can't be build without also building web engine (or at least that seems to be the case for me atm). I am trying to rebuild qt5 with this enabled at the moment.

Is there any specific reason why webengine and declarative are (still) skipped in the port file?

**Comment by ras0219-msft** [comment]
The initial port (https://github.com/Microsoft/vcpkg/pull/219) mentions:

>The webengine, tests, and examples have been excluded as they would significantly increase the build time and storage requirements (it already needs like ~3GB for a dynamic build, and ~7GB for a static build). ... The webengine is also known to be very fickle during builds.

I don't personally know the incremental cost to enable webengine, but one of the improvements we need to make in general is supporting optional features.

For now, the unfortunately best path would be to edit the qt5 portfile locally as you're doing to reenable those pieces. If you could share some results about how much more painful it is to enable those pieces, that would really help to know if we can enable them by default. Even if we can't enable them by default, we'd be happy to accept modifications to the qt5 port that add comments explaining how to enable declarative (specifically that you need to reenable webengine as well) to help out other users like yourself.

**Comment by japj** [comment]
Thanks for the link to the initial port, it is indeed a big build and takes a fair amount of time to verify.
Unfortunately enabling both resulted in a build failure.

I don't know if anyone is already working on #585 , I read that qt configuration was rewritten in 5.8 so it might be easier to first update the port and then see if this is still a problem.

**Comment by nagua** [comment]
I have at least a  `x64-windows` compiling qt5.8 port file. I have also enabled declarative. But I also have the -skip webengine part in the configure part. So I have no idea if the webengine part is build or not.
Also I have not (yet) tested the declarative part.

Also the build time on my i7-6700K is also quiet long. I have not tried to do a static builds yet.
Feel free to use it. Or even bring it to a level, so it can be merged.

I tried a simple test qt5 test app. That worked in Release mode. But I could not get it to run in debug mode, because it could not find the windows plugin (I added the plugins folder to the working dir and it was not working).

So I suspect there is more work to do. But it could be used as a first step for a qt5.8 build.
As I have not many free time for at least mid of april, feel free to use and extend my initial work.

tl:dr Here is my qt5.8 branch https://github.com/nagua/vcpkg/tree/qt5.8 only tested a bit with x64-windows.

**Comment by mmha** [comment]
I also managed to create a portfile with QtQuick enabled and WebEngine disabled. The problem with Qt 5.8 is that it does not build with Visual Studio 2017 (see [this](https://bugreports.qt.io/browse/QTBUG-57086) bug report; the feature test macros got fixed in 5.9).

Thats why I created a portfile for the Qt 5.9 Alpha. I tested it with the `x86-windows` and `x64-windows` configuration and built a QtQuick application with it. You can find it here: https://github.com/mmha/vcpkg/tree/qt59-alpha

Another problem with the Qt5 Port is that it does not properly deploy all required files into the build directory. Off the top of my head I remember Qt5Svg.dll, Qt5Xml.dll, Qt5PrintSupport.dll and everything related to QtQuick (dlls and QtQuick plugins) being affected.

**Comment by nagua** [comment]
I added the patch to my branch to make qt5.8 compatible with VS2017.

**Comment by tobiaskohlbau** [comment]
I've digged into the code from @nagua and found some issues. As I don't know much about the buildsystem from qt, maybe someone has got an idea:

Qt5.8 builds fine with the additions from @nagua but the problem is that it links against the release libraries from zlib, pcre, libpng and libjpeg within the debug build. This results in an error in copying dependencies for an own qt program (Can't find the zlib.dll within the installed debug folder). I will have a further look if it's possible to tell the build system to link against the debug versions.

Feel free to leave you're comments and suggestions.

**Comment by nagua** [comment]
I also noticed this some time ago. One of my patches is used to patch some of the external libraries to link against. I think one has to change all dependencies in there, before doing a debug build. Also I was not able start a application compiled against the debug build of qt, because it could not find the windows plattform. But I copied the plugins folder to the working dir of the application. I someone knows how to do this, please tell me, or do a pull request.

**Comment by tobiaskohlbau** [comment]
@nagua try to run the windeploy tool the problem is you've to create an qt.conf file for it which points prefix path to debug bin folder. The problem is that for debug and release two different qt.conf files must exist but qt is force loading from qt.conf so it's not possible to pass this file by argument. One solution I came up with is to copy tools directory to another location and change qt.conf file according for debug locations.

This is not an very nice solution but for the moment it works. I'm happy to hear other solutions for this problem.

**Comment by luzpaz** [comment]
@japj any luck on the possibility of getting Musescore 3.0.2 in to vcpkg repo ?

**Comment by NancyLi1013** [comment]
Hi @japj, thanks for posting this issue. Does this issue still exist now? If you want to use qml, you can install qt5-declarative. Please try it again.

**Comment by Neumann-A** [comment]
@JackBoosY: This one should be closed

**Comment by JackBoosY** [comment]
Close this issue via fixed.
